core/shared/src/main/scala/zio/prelude/ZValidation.scala

```diff
@@ -0,0 +1,1442 @@
+
+  /**
+   * Transforms the value of this `ZValidation` with the specified validation
+   * function if it is a success or returns the value unchanged otherwise.
+   */
+  final def flatMap[W1 >: W, E1 >: E, B](f: A => ZValidation[W1, E1, B]): ZValidation[W1, E1, B] =
+    self match {
+      case Failure(w, e) => Failure(w, e)
+      case Success(w, a) =>
+        f(a) match {
+          case Failure(w1, e) => Failure(w ++ w1, e)
+          case Success(w1, b) => Success(w ++ w1, b)
+        }
+    }
+
+  /**
+   * Transforms the value of this `ZValidation` with the specified effectual
+   * function if it is a success or returns the value unchanged otherwise.
+   */
+  final def forEach[F[+_]: IdentityBoth: Covariant, B](f: A => F[B]): F[ZValidation[W, E, B]] =
+    self match {
+      case Failure(w, e) => Failure(w, e).succeed[F]
+      case Success(w, a) => f(a).map(Success(w, _))
+    }
+
+  /**
+   * Folds over the error and success values of this `ZValidation`.
+   */
+  final def fold[B](failure: NonEmptyMultiSet[E] => B, success: A => B): B =
+    self match {
+      case Failure(_, e) => failure(e)
+      case Success(_, a) => success(a)
+    }
+
+  /**
+   * Returns the value of the log.
+   */
+  final def getLog: Chunk[W] =
+    self match {
+      case Failure(w, _) => w
+      case Success(w, _) => w
+    }
+
+  /**
+   * Writes an entry to the log.
+   */
+  final def log[W1 >: W](w1: W1): ZValidation[W1, E, A] =
+    self match {
+      case Failure(w, e) => Failure(w :+ w1, e)
+      case Success(w, a) => Success(w :+ w1, a)
+    }
+
+  /**
+   * Transforms the successful value of this `ZValidation` with the specified
+   * function.
+   */
+  final def map[B](f: A => B): ZValidation[W, E, B] =
+    self match {
+      case Failure(w, e) => Failure(w, e)
+      case Success(w, a) => Success(w, f(a))
+    }
+
+  /**
+   * Transforms the error value of this `ZValidation` with the specified
+   * function.
+   */
+  final def mapError[E2](f: E => E2): ZValidation[W, E2, A] =
+    self match {
+      case Failure(w, e) => Failure(w, e.map(f))
+      case Success(w, a) => Success(w, a)
+    }
+
+  /**
+   * Transforms the log entries of this `ZValidation` with the specified
+   * function.
+   */
+  final def mapLog[W2](f: W => W2): ZValidation[W2, E, A] =
+    self match {
+      case Failure(w, e) => Failure(w.map(f), e)
+      case Success(w, a) => Success(w.map(f), a)
+    }
+
+  /**
+   * Exposes the result of this validation function as either a `Right` with
+   * a success of type `A` or a `Left` with one or more errors of type `E`,
+   * along with the log.
+   */
+  final def runLog[B]: (Chunk[W], Either[NonEmptyMultiSet[E], A]) =
+    self match {
+      case Failure(w, e) => (w, Left(e))
+      case Success(w, a) => (w, Right(a))
+    }
+
+  /**
+   * Transforms this `ZValidation` to an `Either`, discarding the log.
+   */
+  final def toEither[E1 >: E]: Either[NonEmptyMultiSet[E1], A] =
+    fold(Left(_), Right(_))
+
```

Comment:
When use `cats.data.Validated` I often accumulate errors into one exception at the end of validation. Because work with `Either[MyDataValidationException, MyData]` is often simpler than with `Either[NonEmptyMultiSet[String], Mydata]` or `Either[NonEmptyChain[String], Mydata]`.

What about `squash` method?

```scala
def squash[E1](f: String => E1)(implicit D: Debug[E]): Either[E1, A] =
  fold(errors => Left(f(errors.toSet.map(D.debug).mkString("; "))), Right(_))

final case class ValidationErrors(errors: String) extends Exception(s"validation failed with [$errors]")
def squash(implicit D: Debug[E]): Either[ValidationErrors, A] =
  sqash(ValidationErrors(_))
```

Reply:
Yes. Maybe something more generic that let's you squash the errors with a function. Will work on this.

Reply:
I don't if this is what you had in mind, but I've made https://github.com/zio/zio-prelude/pull/574